lib/spack/spack/externals.py

```diff
@@ -0,0 +1,349 @@
+
+    requirements = entry["require"]
+    if not isinstance(requirements, list):
+        requirements = [requirements]
+
+    results = []
+    for requirement in requirements:
+        if not isinstance(requirement, str):
+            continue
+
+        matches = _TARGET_RE.match(requirement)
+        if matches:
+            results.append(matches.group(1))
+
+    if len(results) == 1:
+        return results[0]
+
+    return ""
+
+
+class ExternalSpecAndConfig(NamedTuple):
+    spec: spack.spec.Spec
+    config: ExternalDict
+
+
+class ExternalSpecsParser:
+    def __init__(
+        self,
+        external_dicts: List[ExternalDict],
+        *,
+        complete_node: Callable[[spack.spec.Spec], None] = complete_variants_and_architecture,
+        allow_nonexisting: bool = True,
+    ):
+        """Initializes a class to manage and process external specifications.
+
+        Args:
+            external_dicts: list of ExternalDict objects to provide external specifications.
+            complete_node: a callable that completes a node with missing variants, targets, etc.
+                Defaults to `complete_architecture`.
+            allow_nonexisting: whether to allow non-existing packages. Defaults to True.
+
+        Raises:
+            spack.repo.UnknownPackageError: if a package does not exist,
+                and allow_nonexisting is False.
+        """
+        self.external_dicts = external_dicts
+        self.specs_by_external_id: Dict[str, ExternalSpecAndConfig] = {}
+        self.specs_by_name: Dict[str, List[ExternalSpecAndConfig]] = {}
+        self.nodes: List[spack.spec.Spec] = []
+        self.allow_nonexisting = allow_nonexisting
+        # Fill the data structures above (can be done lazily)
+        self.complete_node = complete_node
+        self._parse()
+
+    def _parse(self) -> None:
+        for external_dict in self.external_dicts:
+            try:
+                node = node_from_dict(external_dict)
+            except spack.spec.UnsatisfiableArchitectureSpecError:
+                warnings.warn(
+                    f"cannot constrain external spec '{external_dict['spec']}' with target "
+                    f"'{external_dict['required_target']}'. This spec will not be considered "
+                    f"during concretization."
+                )
+                continue
+            except ExternalSpecError as e:
+                warnings.warn(f"{e} Fix your packages.yaml configuration.")
+                continue
+
+            package_exists = spack.repo.PATH.exists(node.name)
+
+            # If we allow non-existing packages, just continue
+            if not package_exists and self.allow_nonexisting:
+                continue
+
+            if not package_exists and not self.allow_nonexisting:
+                raise spack.repo.UnknownPackageError(node.name, repo=spack.repo.PATH)
+
+            if not package_exists:
+                raise ValueError(f"Package '{node.name}' does not exist")
```

Comment:
Last condition here is unreachable.

```suggestion
            if not package_exists:
                if self.allow_nonexisting:
                    continue
                raise spack.repo.UnknownPackageError(node.name, repo=spack.repo.PATH)
```